cmd/wallet/deposits.go

```diff
@@ -0,0 +1,189 @@
+package walletcmd
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"sort"
+	"text/tabwriter"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/keep-network/keep-core/pkg/bitcoin"
+	"github.com/keep-network/keep-core/pkg/tbtc"
+)
+
+type depositEntry struct {
+	wallet [20]byte
+
+	depositKey      string
+	revealedAtBlock uint64
+	isSwept         bool
+
+	fundingTransactionHash        bitcoin.Hash
+	fundingTransactionOutputIndex uint32
+	amountBtc                     float64
+}
+
+// ListDeposits gets deposits from the chain.
+func ListDeposits(
+	tbtcChain tbtc.Chain,
+	walletStr string,
+	hideSwept bool,
+	sortByAmount bool,
+) error {
+	deposits, err := getDeposits(tbtcChain, walletStr)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get deposits: [%w]",
+			err,
+		)
+	}
+
+	if len(deposits) == 0 {
+		return fmt.Errorf("no deposits found")
+	}
+
+	// Filter
+	if hideSwept {
+		deposits = removeSwept(deposits)
+	}
+
+	// Order
+	sort.SliceStable(deposits, func(i, j int) bool {
+		return deposits[i].revealedAtBlock > deposits[j].revealedAtBlock
+	})
+
+	if sortByAmount {
+		sort.SliceStable(deposits, func(i, j int) bool {
+			return deposits[i].amountBtc < deposits[j].amountBtc
+		})
+	}
+
+	// Print
+	if err := printTable(deposits); err != nil {
+		return fmt.Errorf("failed to print deposits table: %v", err)
+	}
+
+	return nil
+}
+
+func removeSwept(deposits []depositEntry) []depositEntry {
+	result := []depositEntry{}
+	for _, deposit := range deposits {
+		if deposit.isSwept {
+			continue
+		}
+		result = append(result, deposit)
+	}
+	return result
+}
+
+func getDeposits(tbtcChain tbtc.Chain, walletStr string) ([]depositEntry, error) {
+	logger.Infof("reading deposits from chain...")
+
+	filter := &tbtc.DepositRevealedEventFilter{}
+	if len(walletStr) > 0 {
+		walletPublicKeyHash, err := hexToWalletPublicKeyHash(walletStr)
+		if err != nil {
+			return []depositEntry{}, fmt.Errorf("failed to extract wallet public key hash: %v", err)
+		}
+
+		filter.WalletPublicKeyHash = [][20]byte{walletPublicKeyHash}
+	}
+
+	depositRevealedEvent, err := tbtcChain.PastDepositRevealedEvents(filter)
```

Comment:
Did you manage to test it on mainnet with the oldest wallet? Did you try calling it against geth server which is not Infura/Alchemy? If not, can you please perform such a test? We had a [discussion](https://github.com/keep-network/keep-core/pull/3521/files#r1173592721) with @lukasz-zimnoch on how long it takes to retrieve old events without passing a block range. I expect Alchemy and Infura may have some fancier cache/indexing but does it work fast enough with a standard geth? Knowing this would help us determine how fancy we need to be with calls like that and this simple CLI command is a perfect opportunity.

Reply:
### Get All Deposits (no filter):

**Mainnet:**
Alchemy: 0.205702 seconds
Infura: 0.523039 seconds
Geth: 265.665695 seconds

### Get Wallet's Deposits (with wallet filter):

**Mainnet:**
Alchemy: 0.186728 seconds
Infura: 0.271587 seconds
Geth: 114.994290 seconds